src/server/game/Scenarios/Scenario.cpp

```diff
@@ -0,0 +1,307 @@
+}
+
+void Scenario::SetStep(ScenarioStepEntry const* step)
+{
+    _currentstep = step;
+    if (step)
+        SetStepState(step, SCENARIO_STEP_IN_PROGRESS);
+
+    WorldPackets::Scenario::ScenarioState scenarioState;
+    BuildScenarioState(&scenarioState);
+    SendPacket(scenarioState.Write());
+}
+
+void Scenario::OnPlayerEnter(Player* player)
+{
+    m_players.insert(player->GetGUID());
+    SendScenarioState(player);
+}
+
+void Scenario::OnPlayerExit(Player* player)
+{
+    m_players.erase(player->GetGUID());
+    SendBootPlayer(player);
+}
+
+bool Scenario::IsComplete()
+{
+    for (auto step : _data->Steps)
+    {
+        if (step.second->IsBonusObjective())
+            continue;
+
+        if (GetStepState(step.second) != SCENARIO_STEP_DONE)
+            return false;
+    }
+
+    return true;
+}
+
+ScenarioStepState Scenario::GetStepState(ScenarioStepEntry const* step)
+{
+    std::map<ScenarioStepEntry const*, ScenarioStepState>::const_iterator itr = _stepStates.find(step);
+    if (itr == _stepStates.end())
+        return ScenarioStepState(NULL);
+
+    return itr->second;
+}
+
+void Scenario::SendCriteriaUpdate(Criteria const * criteria, CriteriaProgress const * progress, uint32 timeElapsed, bool timedCompleted) const
+{
+    WorldPackets::Scenario::ScenarioProgressUpdate progressUpdate;
+    WorldPackets::Achievement::CriteriaProgress criteriaProgress;
+    criteriaProgress.Id = criteria->ID;
+    criteriaProgress.Quantity = progress->Counter;
+    criteriaProgress.Player = progress->PlayerGUID;
+    criteriaProgress.Date = progress->Date;
+    if (criteria->Entry->StartTimer)
+        criteriaProgress.Flags = timedCompleted ? 1 : 0;
+
+    criteriaProgress.TimeFromStart = timeElapsed;
+    criteriaProgress.TimeFromCreate = 0;
+
+    progressUpdate.CriteriaProgress = criteriaProgress;
+    SendPacket(progressUpdate.Write());
+}
+
+bool Scenario::CanUpdateCriteriaTree(Criteria const * /*criteria*/, CriteriaTree const * tree, Player * /*referencePlayer*/) const
+{
+    ScenarioStepEntry const* step = tree->ScenarioStep;
+    if (!step)
+        return false;
+
+    if (step->ScenarioID != _data->Entry->ID)
+        return false;
+
+    ScenarioStepEntry const* currentStep = GetStep();
+    if (!currentStep)
+        return false;
+
+    if (step->IsBonusObjective())
+        return true;
+
+    return currentStep == step;
+}
+
+bool Scenario::CanCompleteCriteriaTree(CriteriaTree const* tree)
+{
+    ScenarioStepEntry const* step = tree->ScenarioStep;
+    if (!step)
+        return false;
+
+    if (step->ScenarioID != _data->Entry->ID)
+        return false;
+
+    if (step->IsBonusObjective())
+        return !IsComplete();
+
+    if (step != GetStep())
+        return false;
+
+    return true;
+}
+
+void Scenario::CompletedCriteriaTree(CriteriaTree const * tree, Player * referencePlayer)
```

Comment:
Thanks tkrokli, I'll see if I can find more unused parameters other than that for the next commit :)